flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java

```diff
@@ -110,14 +124,20 @@ void onConsumedPartition(ResultPartition partition) {
 		LOG.debug("Received consume notification from {}.", partition);
 
 		synchronized (registeredPartitions) {
-			final ResultPartition previous = registeredPartitions.remove(partition.getPartitionId());
-			// Release the partition if it was successfully removed
-			if (partition == previous) {
-				partition.release();
-				ResultPartitionID partitionId = partition.getPartitionId();
-				LOG.debug("Released partition {} produced by {}.",
-					partitionId.getPartitionId(), partitionId.getProducerId());
-			}
+			registeredPartitions.computeIfPresent(
+				partition.getPartitionId(),
+				(resultPartitionID, resultPartition) -> {
+					if (partition == resultPartition) {
+						if (partition.isManagedExternally() && !isReleaseExternallyManagedPartitionsOnConsumption) {
```

Comment:
I have another option for this. ATM both `ResultPartition` and `ResultPartitionManager` need check the conditions of releasing partitions consumed, then the logic seems difficult to trace. 
We could make `ResultPartitionManager` not aware of this. The `isReleaseExternallyManagedPartitionsOnConsumption` could be merged with `partition.isManagedExternally()` to generate one final boolean tag in `ResultPartition`. Then the `ResultPartition` checks this tag to decide whether to call `partitionManager.onConsumedPartition` while one sub partition is consumed. The logic in `ResultPartitionManager` would be simple:

- Based on consumption release: The behavior is the same as before after calling `ResultPartitionManger#onConsumedPartition`, remove from map and release directly.

- Based on external release: via `ResultPartitionManager#releasePartition`.

And we only have one boolean tag in `ResultPartition` to handle the external check. WDYT? 
This suggestion is not relevant with pin issue.

Reply:
... I never thought about combining isManagedExternally and the configuration flag. That's a great idea.

Reply:
Implemented.